src/tasks/processBulkUpload.ts

```diff
@@ -72,6 +99,204 @@ const downloadS3ObjectToTemporaryStorage = async (key: string): Promise<FileResu
   return temporaryFile;
 };
 
+const loadShortCodesFromBulkUploadCsv = async (csvPath: string): Promise<string[]> => {
+  let shortCodes: string[] = [];
+  let hasLoadedShortCodes = false;
+  const parser = fs.createReadStream(csvPath).pipe(
+    // Loading the entire CSV is a waste, but the `to` option is currently broken
+    // See https://github.com/adaltas/node-csv/issues/410
+    parse(),
+  );
+  await parser.forEach(async (record: string[]) => {
+    if (!hasLoadedShortCodes) {
+      shortCodes = record;
+      hasLoadedShortCodes = true;
+    }
+  });
+  return shortCodes ?? [];
+};
+
+const assertShortCodesAreValid = async (shortCodes: string[]): Promise<void> => {
+  const baseFields = await loadBaseFields();
+  if (!shortCodes.find(
+    (shortCode) => shortCode === RequiredBulkUploadFields.PROPOSAL_SUBMITTER_EMAIL.valueOf(),
+  )) {
+    throw new Error(`${RequiredBulkUploadFields.PROPOSAL_SUBMITTER_EMAIL} is a required field.`);
+  }
+  shortCodes.forEach((shortCode) => {
+    const baseField = baseFields.find(
+      (baseFieldCandidate) => baseFieldCandidate.shortCode === shortCode,
+    );
+    if (baseField === undefined) {
+      throw new Error(`${shortCode} is not a valid BaseField short code.`);
+    }
+  });
+};
+
+const assertCsvContainsValidShortCodes = async (csvPath: string): Promise<void> => {
+  const shortCodes = await loadShortCodesFromBulkUploadCsv(csvPath);
+  if (shortCodes.length === 0) {
+    throw new Error('No short codes detected in the first row of the CSV');
+  }
+  await assertShortCodesAreValid(shortCodes);
+};
+
+const assertCsvContainsRowsOfEqualLength = async (csvPath: string): Promise<void> => {
+  const csvReadStream = fs.createReadStream(csvPath);
+  const parser = parse();
+  parser.on('readable', () => {
+    while ((parser.read()) !== null) {
+      // Iterate through the data -- an error will be thrown if
+      // any columns have a different number of fields
+      // see https://csv.js.org/parse/options/relax_column_count/
+    }
+  });
+  csvReadStream.pipe(parser);
+  await finished(parser);
+};
+
+const assertBulkUploadCsvIsValid = async (csvPath: string): Promise<void> => {
+  await assertCsvContainsValidShortCodes(csvPath);
+  await assertCsvContainsRowsOfEqualLength(csvPath);
+};
+
+export const getApplicantEmailFieldIndexForBulkUpload = async (
+  csvPath: string,
+): Promise<number> => {
+  const shortCodes = await loadShortCodesFromBulkUploadCsv(csvPath);
+  return shortCodes.findIndex(
+    (shortCode) => shortCode === RequiredBulkUploadFields.PROPOSAL_SUBMITTER_EMAIL.valueOf(),
+  );
+};
+
+const createOpportunityForBulkUpload = async (): Promise<Opportunity> => {
+  const currentDate = new Date();
+  const result = await db.sql<Opportunity>('opportunities.insertOne', {
+    title: `Bulk Upload (${currentDate.getMonth()}/${currentDate.getDate()}/${currentDate.getFullYear()})`,
```

Comment:
Downside of bulk upload ID is ID is sequential across users and that might seem confusing to make so visible / there might be gaps.  I also don't think ID is useful to an end user, but timestamp is.  I agree the upload should be the time used rather than the processing time.  (I'm OK with `yyyy-mm-dd` -- the world is better for it!)